File: go/armstrong-numbers/armstrong_numbers.go
// Package armstrong determines if a number is an Armstrong number.
package armstrong

// Using strconv.Itoa is one way to do this, but where's the fun
// in that?

// IsNumber tests a given number for Armstrongness.
func IsNumber(n int) bool {
	// normalize
	if n < 0 {
		n = -n
	}

	// find starting power of 10 for
	// digitizing
	power, prior, count := 1, 1, 1
	for power <= n {
		prior = power
		power *= 10
		count++
	}
	power = prior

	// extract digits, from the most to the least significant.
	digits := make([]int, 0, count)
	temp := n
	for power > 1 {
		d := temp / power
		temp %= power
		power /= 10
		digits = append(digits, d)
	}
	digits = append(digits, temp)

	// armstrong numbers are those that is the sum of its
	// digits each raised to the power of the number of
	// digits
	fraise := func(b, p int) int {
		if p == 0 {
			return 1
		}
		n := b
		for r := 1; r < p; r++ {
			n *= b
		}
		return n
	}
	sum := 0
	for _, d := range digits {
		sum += fraise(d, len(digits))
	}

	return sum == n
}

File: c/armstrong-numbers/armstrong_numbers.c
// exercism armstrong numbers
// t.brumley june 2022

#include <stdio.h> 
#include <stdlib.h>
#include <errno.h>
#include <math.h>
#include <string.h>

#include "armstrong_numbers.h"

// is a given integer an armstrong number?
//
// an armstrong number is one where the sum of its digits each raised
// to the power of the number of digits equals the number.
//
// 9 -> 9^1 = 9 ** is **
// 10 -> 1^2 + 0^2 = 1 ** not **
// 4 -> 4^1 = 4 ** is **

bool is_armstrong_number(int candidate)
{
    // quick guards for some easy cases
    if (candidate >= 0 && candidate < 10)
        return true;
    if (candidate > 9 && candidate < 100)
        return false;

    // the exercise instructions are to use standard library support. i'd
    // prefer to avoid the bloat, but i'll follow the instructions.

    // convert to string for easy access to digits
    // SB_LENGTH includes trailing null byte
#define SB_LENGTH 32
    char sb[SB_LENGTH];
    snprintf(sb, SB_LENGTH, "%d", candidate);

    // sum each digit raised to the number of digits
    int sum = 0;
    double pwr = strlen(sb);
    for(char* p = sb; *p != '\0'; p++) {
        int d = *p - '0';
        sum += (int)(pow((double)d, pwr));
    }

    // if sum and original value match, the eagle has landed
    return sum == candidate;

}
